### PR TITLE
4553 - Refactor RenkoConsolidator

### LINKED ISSUE #4553: RenkoConsolidator Division By Zero
#### Expected Behavior
Do not allow setting RenkoConsolidator.BarSize to zero.

#### Actual Behavior
It's possible to set `RenkoConsolidator.BarSize`. It causes unhandled `DivideByZeroException` in `RenkoConsolidator.UpdateClassic`

#### Potential Solution
Thrown `ArgumentException` in `RenkoConsolidator` construction.

#### Reproducing the Problem
```python
renkoConsolidator = RenkoConsolidator(round(self.atr.Current.Value,4))
```
where 
`round(self.atr.Current.Value,4)` is zero.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by aarjaneiro**
Refactored:
```
                         ↗ RenkoConsolidator
 RenkoConsolidator
                         ↘ WickedRenkoConsolidator
```
#### Description

Created a base class from which both inherit. 

#### Related Issue

Closes #4553 
#### Motivation and Context

Calling RenkoType.Wicked separated initialization and update conditions, better to have separate classes altogether.

#### Requires Documentation Change

No
#### How Has This Been Tested?

Tests as implemented for the Renko Consolidator
#### Types of changes

- [x ] Bug fix (non-breaking change which fixes an issue)
- [x ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x ] My code follows the code style of this project.
- [x ] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x ] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x ] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Data/Consolidators/BaseRenkoConsolidator.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        /// </summary>
+        public Type OutputType => typeof(RenkoBar);
+
+
+        /// <summary>
+        /// Updates this consolidator with the specified data
+        /// </summary>
+        /// <param name="data">The new data for the consolidator</param>
+        public abstract void Update(IBaseData data);
+
```
Comment (C-SELLERS) [inline Common/Data/Consolidators/BaseRenkoConsolidator.cs:171]: nit - extra new lines

File: Common/Data/Consolidators/BaseRenkoConsolidator.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+using System;
+using Python.Runtime;
+using QuantConnect.Data.Market;
+
+namespace QuantConnect.Data.Consolidators
+{
+    /// <summary>
+    /// The base class for the RenkoConsolidator class between <see cref="RenkoType"/>s.
+    /// </summary>
+    public abstract class BaseRenkoConsolidator : IDataConsolidator
```
Comment (C-SELLERS) [inline Common/Data/Consolidators/BaseRenkoConsolidator.cs:25]: Nice!

File: Common/Data/Consolidators/RenkoConsolidator.cs [hunk 3]
```diff
@@ -122,264 +64,51 @@ public RenkoConsolidator(decimal barSize, Func<IBaseData, decimal> selector, Fun
...
-                        _highRate = limit;
-
-                        Falling(data);
-                    }
-                }
-            }
-        }
-
-        private void UpdateClassic(IBaseData data)
+        public override void Update(IBaseData data)
```
Comment (C-SELLERS) [inline Common/Data/Consolidators/RenkoConsolidator.cs:81]: 👍🏽

File: Common/Data/Consolidators/BaseRenkoConsolidator.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            else
+            {
+                VolumeSelector = (x => 0);
+            }
+        }
+
+        /// <summary>
+        /// Gets the kind of the bar
+        /// </summary>
+        public RenkoType Type { get; }
```
Comment (C-SELLERS) [inline Common/Data/Consolidators/BaseRenkoConsolidator.cs:136]: nit - members above the constructor ordered by accessibility

**Review by C-SELLERS — CHANGES_REQUESTED** [review]
Looks good to me! Just needs a once over on the nit details.

File: Common/Data/Consolidators/BaseRenkoConsolidator.cs (hunk 4 above)
Comment (aarjaneiro) [reply Common/Data/Consolidators/BaseRenkoConsolidator.cs:124]: Ah, it seems I was following the wrong convention! I will reorder as in https://stackoverflow.com/questions/150479/order-of-items-in-classes-fields-properties-constructors-methods

**Review by C-SELLERS — APPROVED** [review]

File: Common/Data/Consolidators/BaseRenkoConsolidator.cs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RenkoConsolidator" /> class.
+        /// </summary>
+        /// <param name="barSize">The size of each bar in units of the value produced by <paramref name="selector"/></param>
+        /// <param name="selector">Extracts the value from a data instance to be formed into a <see cref="RenkoBar"/>. The default
+        /// value is (x => x.Value) the <see cref="IBaseData.Value"/> property on <see cref="IBaseData"/></param>
+        /// <param name="volumeSelector">Extracts the volume from a data instance. The default value is null which does
+        /// not aggregate volume per bar.</param>
+        /// <param name="evenBars">When true bar open/close will be a multiple of the barSize</param>
+        public BaseRenkoConsolidator(decimal barSize, PyObject selector, PyObject volumeSelector = null, bool evenBars = true)
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/BaseRenkoConsolidator.cs:89]: nit - `public` -> `protected`

**Review by Martin-Molinero — COMMENTED** [review]
Thanks for the PR!
Leaving a couple of requests, basically let's abstract the base Renko consolidator a bit more, in such a way he only holds shared logic/properties/methods and each implementation owns its particular details/behavior

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
See previous comments 👍

**Comment by Martin-Molinero** [comment]
Believe stats have changed for the `WarmupIndicatorRegressionAlgorithm` & `RenkoConsolidatorAlgorithm` with this branch? Should review _why_ and revert the change causing it, if appropriate, wouldn't expect algorithm stats changed with this PR

File: Common/Data/Consolidators/BaseRenkoConsolidator.cs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+{
+    /// <summary>
+    /// The base class for the RenkoConsolidator class between <see cref="RenkoType"/>s.
+    /// </summary>
+    public abstract class BaseRenkoConsolidator : IDataConsolidator
+    {
+        internal readonly Func<IBaseData, decimal> Selector;
+        internal readonly Func<IBaseData, decimal> VolumeSelector;
+        internal readonly bool EvenBars;
+        internal RenkoBar CurrentBar;
```
[the remark this answers is not in the document]
Comment (aarjaneiro) [reply Common/Data/Consolidators/BaseRenkoConsolidator.cs:30]: Good catch on the Selector! I will change the access modifiers and keep to the styling as well.

**Comment by aarjaneiro** [comment]
> See previous comments +1

Ended up removing the base class as little really ends up in common after refactoring (really only Renko type as a property and the epsilon checking (void) function could have been generalized in a base class).

File: Algorithm.CSharp/WarmupIndicatorRegressionAlgorithm.cs [hunk 7]
```diff
@@ -17,6 +17,7 @@
 using System.Collections.Generic;
 using QuantConnect.Data;
 using QuantConnect.Data.Consolidators;
+using QuantConnect.Data.Market;
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/WarmupIndicatorRegressionAlgorithm.cs:20]: nit - unrequired new import

File: Common/Data/Consolidators/RenkoConsolidator.cs [hunk 8]
```diff
@@ -81,12 +66,11 @@ public RenkoConsolidator(decimal barSize, RenkoType type)
...
         public RenkoConsolidator(decimal barSize, bool evenBars = true)
         {
+            EpsilonCheck(barSize);
             _barSize = barSize;
             _selector = x => x.Value;
-            _volumeSelector = x => 0;
             _evenBars = evenBars;
-
-            Type = RenkoType.Classic;
+            _volumeSelector = x => 0;
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/RenkoConsolidator.cs:73]: Believe this constructor could call the one bellow `: this(barSize, x => x.Value, x => 0, evenBars)`

File: Common/Data/Consolidators/RenkoConsolidator.cs [hunk 9]
```diff
@@ -122,205 +105,63 @@ public RenkoConsolidator(decimal barSize, Func<IBaseData, decimal> selector, Fun
...
-                OnDataConsolidated(wicko);
-
-                _openOn = _closeOn;
-                _openRate = limit;
-                _lowRate = limit;
-            }
-        }
-
-        private void Falling(IBaseData data)
+        event DataConsolidatedHandler IDataConsolidator.DataConsolidated
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/RenkoConsolidator.cs:154]: nit - would be nice to move `IDataConsolidator.DataConsolidated` & `event EventHandler<RenkoBar> DataConsolidated` above the constructors

File: Common/Data/Consolidators/RenkoConsolidator.cs [hunk 10]
```diff
@@ -360,27 +202,36 @@ public void Scan(DateTime currentLocalTime)
...
+        {
+            DataConsolidated = null;
+            _dataConsolidatedHandler = null;
+        }
+
+        private static void EpsilonCheck(decimal barSize)
+        {
+            if (barSize < Extensions.GetDecimalEpsilon())
+            {
+                throw new ArgumentOutOfRangeException(nameof(barSize),
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/RenkoConsolidator.cs:217]: Cool!

File: Common/Data/Consolidators/WickedRenkoConsolidator.cs [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        private DateTime _openOn;
+        private decimal _openRate;
+        private decimal _barSize;
+        private DataConsolidatedHandler _dataConsolidatedHandler;
+        private RenkoBar _currentBar;
+        private IBaseData _consolidated;
+
+        // Used for unit tests
+        internal RenkoBar OpenRenkoBar =>
+            new RenkoBar(null, _openOn, _closeOn, _barSize, _openRate, _highRate, _lowRate, _closeRate);
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/WickedRenkoConsolidator.cs:42]: Could we directly remove `OpenRenkoBar` and use `WorkingData` instead?

File: Common/Data/Consolidators/WickedRenkoConsolidator.cs [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+        /// <summary>
+        /// Event handler that fires when a new piece of data is produced
+        /// </summary>
+        public event EventHandler<RenkoBar> DataConsolidated;
+
+        /// <summary>
+        /// Event handler that fires when a new piece of data is produced
+        /// </summary>
+        event DataConsolidatedHandler IDataConsolidator.DataConsolidated
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/WickedRenkoConsolidator.cs:92]: nit - The same comment as the classic renko, let's move these last 2 guys above the constructor

File: Tests/Common/Data/WickedRenkoConsolidatorTests.cs [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+using QuantConnect.Indicators;
+using System.Collections.Generic;
+
+namespace QuantConnect.Tests.Common.Data
+{
+    [TestFixture]
+    public class WickedRenkoConsolidatorTests
+    {
+        [Test]
+        public void WickedOutputTypeIsRenkoBar()
```
Comment (Martin-Molinero) [inline Tests/Common/Data/WickedRenkoConsolidatorTests.cs:29]: Nice! 💯

File: Tests/Common/Data/RenkoConsolidatorTests.cs [hunk 14]
```diff
@@ -18,11 +18,9 @@
 using QuantConnect.Data.Consolidators;
 using QuantConnect.Data.Market;
 using QuantConnect.Indicators;
-using System.Collections.Generic;
 
 namespace QuantConnect.Tests.Common.Data
 {
-    [TestFixture]
```
Comment (Martin-Molinero) [inline Tests/Common/Data/RenkoConsolidatorTests.cs:25]: Wondering why remove `[TestFixture]` here?

File: Common/Data/Consolidators/RenkoConsolidator.cs [hunk 15]
```diff
@@ -24,53 +24,38 @@ namespace QuantConnect.Data.Consolidators
...
+        /// </summary>
+        public Type InputType => typeof(IBaseData);
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="RenkoConsolidator"/> class using the specified <paramref name="barSize"/>.
+        /// Gets <see cref="RenkoBar"/> which is the type emitted in the <see cref="IDataConsolidator.DataConsolidated"/> event.
         /// </summary>
-        /// <param name="barSize">The constant value size of each bar</param>
-        /// <param name="type">The RenkoType of the bar</param>
-        public RenkoConsolidator(decimal barSize, RenkoType type)
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/RenkoConsolidator.cs:64]: For a happier user experience suggest we keep this constructor, flag it as obsolete, for example: `[Obsolete("Please use WickedRenkoConsolidator type")]` and make it thrown an exception with a similar message

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Sweet! Almost there, leaving a few minor comments & the regression tests `WarmupIndicatorRegressionAlgorithm` & `RenkoConsolidatorAlgorithm (C#/Py)` are still failing in this branch, should review why & address/fix

**Comment by Martin-Molinero** [comment]
If you're up to it, it would be great if after this PR you took a stab at this renko related bug https://github.com/QuantConnect/Lean/issues/2853

File: Common/Data/Consolidators/WickedRenkoConsolidator.cs (hunk 11 above)
Comment (aarjaneiro) [reply Common/Data/Consolidators/WickedRenkoConsolidator.cs:42]: This is interesting. From what I see, this is a means of bypassing the fact that for the end user, assignment of `_currentBar` occurs upon `OnDataConsolidated`. For a property like `WorkingData`, I reckon its best to change each of `_openOn, _closeOn, _barSize, _openRate, _highRate, _lowRate, _closeRate` to `internal` than to determine what is returned based on whether or not the caller is a test. Doing this and then adding a method in WickedConsolidatorTests which returns the same value as `OpenRenkoBar` can allow removal of `OpenRenkoBar`.

File: Common/Data/Consolidators/RenkoConsolidator.cs (hunk 8 above)
Comment (aarjaneiro) [reply Common/Data/Consolidators/RenkoConsolidator.cs:87]: I tried that, though it seems to cause a cyclic constructor call

**Comment by aarjaneiro** [comment]
> Sweet! Almost there, leaving a few minor comments & the regression tests `WarmupIndicatorRegressionAlgorithm` & `RenkoConsolidatorAlgorithm (C#/Py)` are still failing in this branch, should review why & address/fix

My next commit will focus on this now that the main "structure" provided by this refactor is essentially there.

File: Algorithm.CSharp/RenkoConsolidatorAlgorithm.cs [hunk 16]
```diff
@@ -25,6 +25,7 @@ namespace QuantConnect.Algorithm.CSharp
     /// </summary>
     /// <meta name="tag" content="renko" />
     /// <meta name="tag" content="indicators" />
+    /// <meta name="tag" content="wicked"/>
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/RenkoConsolidatorAlgorithm.cs:28]: nit - unrequired change?

File: Common/Data/Consolidators/RenkoConsolidator.cs (hunk 8 above)
Comment (Martin-Molinero) [reply Common/Data/Consolidators/RenkoConsolidator.cs:87]: Please try `: this(barSize, x => x.Value, x => 0, evenBars)` should work, the other constructor isn't calling no other 👍

File: Common/Data/Consolidators/RenkoConsolidator.cs [hunk 17]
```diff
@@ -98,19 +96,34 @@ public RenkoConsolidator(decimal barSize, bool evenBars = true)
...
+        /// <param name="type"></param>
+        [Obsolete("Please use the WickedRenkoConsolidator if RenkoType is not Classic")]
+        public RenkoConsolidator(decimal barSize, RenkoType type)
+        {
+            if (type != RenkoType.Classic)
+            {
+                throw new ArgumentException("Please use the WickedRenkoConsolidator type if RenkoType is not Classic");
+            }
+
+            _barSize = barSize;
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/RenkoConsolidator.cs:126]: nit - similar let's call `: this(barSize, true)` instead

File: Common/Data/Consolidators/WickedRenkoConsolidator.cs [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+    /// </summary>
+    public class WickedRenkoConsolidator : IDataConsolidator
+    {
+        internal DateTime CloseOn;
+        internal decimal CloseRate;
+        internal decimal HighRate;
+        internal decimal LowRate;
+        internal DateTime OpenOn;
+        internal decimal OpenRate;
+        internal decimal BarSize;
```
Comment (Martin-Molinero) [inline Common/Data/Consolidators/WickedRenkoConsolidator.cs:33]: Let's please make these protected and create a test class `WickedRenkoConsolidatorTest` exposing them so they can be asserted on 👍

Comment (aarjaneiro) [reply Common/Data/Consolidators/WickedRenkoConsolidator.cs:33]: Alternatively, changing them to protected internals will expose them without needing to re-add something of the form of the removed `OpenRenkoBar`. Would that be preferable?

**Review by Martin-Molinero — APPROVED** [review]

File: Algorithm.CSharp/RenkoConsolidatorAlgorithm.cs (hunk 16 above)
Comment (jaredbroad) [reply Algorithm.CSharp/RenkoConsolidatorAlgorithm.cs]: This helps our documentation search for the types

**Pull request merged**